### PR TITLE
1221 - Tapestry 3.0 - Allow standalone nodes

### LINKED ISSUE #1221: Allow standalone nodes
It should be possible for nodes without any connections to exist by themselves in the Tapestry.

To achieve this, the notion of a "root node" for the Tapestry should be abandoned and all code referring to or using the root node should be removed. Also the checks for whether a node has any connections before showing them should also be removed.

### LINKED ISSUE #1223: "Permission denied warning" when deleting rejected node
Branch: found on #1191 , but probably also exists in master.

Steps to reproduce:
1. Submit a node for review as a subscriber.
2. Reject the node as an administrator.
3. As a subscriber, try to delete the rejected node - it will show a "You don't have permission" alert, but still delete the node.

### PR DESCRIPTION
**Pull request by legendword**
## Changes
* Make `rootId` field of the tapestry object **optional**, so that it is merely used to determine the initial node selected upon viewing the Tapestry for the first time. **Root node should be called "default node" from now on, since it reflects which node gets selected when viewing the tapestry for the first time.**
* Allow deleting any node and any link.
* Whenever a node is added to an **empty** tapestry, that node will become the default node. If the default node has been changed to a node different than the first node added (which is not possible with this PR - but this PR opens up the possibility for a future PR to implement the ability for authors to set a node as the default node), and the default node is deleted, then the node selected for a first-time viewer of the tapestry will fall back to the first node in the tapestry by the order of creation until otherwise overwritten by setting an explicit `rootId` value.
* Allow any user to create a new node on an empty tapestry (for non-admins, they will have to submit the node for review).

### New draft node behaviours
* Administrators should be able to directly publish any draft node, including a standalone draft node, and a draft node that is only attached to other draft nodes.
* Updated draft\* node permissions for author / reviewer (\*Note: accepted draft nodes are PUBLISHED nodes, so they are not considered as draft nodes here):
  * If node is not submitted for review:
    * Allow all actions except "add" for original author
    * Allow all actions except "edit" for reviewer if node is rejected and showRejected is true
  * If node is submitted for review:
    * Only allow "read" for original author
    * Allow all actions except "edit" for reviewer
* Users can now create a standalone draft node via a "hack": first they can add a draft child node to a published node, then they can remove the link from the published node to the draft node making their draft node standalone. *[need to verify if this is expected behaviour]*

### Bug fixes and improvements
* Bug fix: the minimap is showing even when the Tapestry is empty.
* Bug fix: links associated with a node are not being deleted in the backend when the node is deleted. This is a small but significant bug due to an associative array not being cast to a PHP object. The bug does not affect any visuals because links to nonexistent nodes are filtered out by the backend & frontend - it is also the reason why this bug was not found until this PR.
* Improvement: the error message and status code of a TapestryError in the backend can be overridden by explicitly specifying them; if not overridden, the values in the TapestryError's ERRORS array will be used.

## Screenshot
image

## Issue Linkage
Closes #1221 

Closes #1223 
## PR Dependency
Depends on: #1276 
## Automated Testing
* N/A
* (unsure if we need some automated testing for this)

### CODE REVIEW

**Comment by legendword** [comment]
image

A bug with creating draft nodes:

1. create a node
2. add a draft child node to it
3. delete the parent node
4. submit the child node for review
5. child node cannot be accepted since the parent node is now gone

Is this the expected behaviour?

**Comment by legendword** [comment]
While debugging I noticed this piece of code in the Sidebar component, where if el is found it calls `console.error`. Is this a mistake?
https://github.com/tapestry-tool/tapestry-wp/blob/74457c3a3e3a8eab39e3b3a774ec4330dff01e4c/templates/vue/src/components/Sidebar/index.vue#L251-L265

**Comment by legendword** [comment]
Bug: add node in empty tapestry results in a link to itself (happens after deleting all existing nodes and adding a first node again)

image

**Comment by legendword** [comment]
**Bug**: sometimes an extra "updateNode" is triggered with the ID of an already deleted node when adding a node in an empty tapestry after deleting all nodes, which causes the backend to throw NodeMetaId invalid error. I have yet to be able to consistently reproduce this bug but it is a critical one that needs fixing.

**Comment by legendword** [comment]
**Bug**: links attached to deleted nodes are still present in the backend (can be seen within the export file of the tapestry)

**Comment by zifgu** [comment]
Tested deleting nodes and links in different arrangements; they are being deleted properly, including when inspected in the dev tools. The special cases of deleting Multi-Content nodes with children or Video nodes with pop-ups also work well.

I haven't reviewed the code just yet, but the available operations are different when a draft node created by an admin has a parent and when it is standalone. I am actually not sure which behaviour is correct, but I would expect both cases to have the same available operations.

I also encountered some "You don't have permission" alerts while trying the new operations - for example:

https://user-images.githubusercontent.com/44929104/181660739-977091b2-ad6e-454f-98c8-bb38bfea13da.mp4

However, this may be related to the issue reported in #1223, so I think we should investigate that first.

Lastly, does the code contain changes from #1212? If so, it would be great if we could somehow merge the two branches before merging in this PR.

**Comment by zifgu** [comment]
[Noting a behaviour that results from this PR]
When an admin rejects a draft node, it disappears from the admin's view but is visible to the person who originally submitted it as a rejected node. The nodes and links to the rejected node still exist and can be seen in the minimap.

However, if the admin deletes everything else in the Tapestry while it still contains rejected nodes, the admin will seean empty view instead of the "Add root node" screen, and will not be able to add any more nodes.

This should no longer be a problem after the "add node" tool is implemented.

File: templates/vue/src/components/TapestryApp.vue [hunk 1]
```diff
@@ -28,9 +28,10 @@ export default {
     return {}
   },
   computed: {
-    ...mapState(["nodes", "links", "selection", "settings", "rootId"]),
+    ...mapState(["nodes", "links", "selection", "settings"]),
+    ...mapGetters(["isEmptyTapestry"]),
     isSidebarOpen() {
-      return Boolean(this.$route.query.sidebar)
+      return Boolean(this.$route.query.sidebar) && !this.isEmptyTapestry
```
Comment (zifgu) [inline templates/vue/src/components/TapestryApp.vue:34]: Just wondering, what is the purpose of adding the condition that the Tapestry not be empty here?

**Review by zifgu — COMMENTED** [review]
The code changes all make sense to me. Once we resolve the difference in available operations that I described above, I can mark this as ready for final review.

**Comment by zifgu** [comment]
> I also encountered some "You don't have permission" alerts while trying the new operations

I have just investigated what might be causing this issue (see #1223 for the details of the bug). Since this branch has the latest changes to node removal, I have just pushed the fix (one line) as 97c7ed2 here rather than creating a separate PR. If this doesn't fix the issue, please let me know!

File: templates/vue/src/components/TapestryMain/index.vue [hunk 2]
```diff
@@ -218,9 +217,6 @@ export default {
   },
   methods: {
     ...mapMutations(["select", "unselect", "clearSelection"]),
-    addRootNode() {
-      this.$root.$emit("add-node", null)
-    },
```
[the remark this answers is not in the document]
Comment (legendword) [reply templates/vue/src/components/TapestryMain/index.vue:223]: I now see what this code is doing... it is apparently publishing the root node as the default node template at the same time as the node modal is opened. I removed it because I thought it was a bit redundant since we already open the node modal for the author to create the first node, and the handleSubmit will be called once the author clicks on the "Publish" button. But there may be a reason why we are immediately submitting the root node which I'm not aware of.

File: templates/vue/src/components/modals/NodeModal/index.vue [hunk 3]
```diff
@@ -748,22 +724,26 @@ export default {
...
           this.$router.push({ path: "/", query: this.$route.query })
         } else if (this.returnRoute) {
           this.$router.push(this.returnRoute)
         } else if (this.keepOpen) {
-          // Switch to edit mode if multi-content just added
           this.$router.push({
             name: names.MODAL,
-            params: { nodeId: this.node.id, type: "edit", tab: "content" },
+            params: {
+              nodeId: this.keepOpen === true ? this.node.id : this.keepOpen,
```
[the remark this answers is not in the document]
Comment (legendword) [reply templates/vue/src/components/modals/NodeModal/index.vue:735]: In `this.$root.$on("add-node", ...)`, this.keepOpen is set to the current node which has the value of true. But when deleting a node and going back to its parent, this.keepOpen is set to the nodeId of the parent node. That is why `keepOpen` can be either `true` or an integer representing the nodeId to go to.

File: templates/vue/src/components/TapestryApp.vue (hunk 1 above)
Comment (legendword) [reply templates/vue/src/components/TapestryApp.vue:34]: On an empty Tapestry, the sidebar is hidden but the main tapestry view still leaves space for the sidebar, resulting in an awkward white space on the right. Adding this just makes the spacing consistent with the actual visibility of the sidebar.

File: templates/vue/src/components/TapestryMain/index.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (zifgu) [reply templates/vue/src/components/TapestryMain/index.vue:223]: Thanks for explaining. That makes sense. I created root nodes many times on this PR and haven't noticed anything incorrect, so deleting it seems reasonable to me as well.

File: templates/vue/src/components/TapestryApp.vue (hunk 1 above)
Comment (zifgu) [reply templates/vue/src/components/TapestryApp.vue:34]: Makes sense now. Thanks for explaining.

**Comment by legendword** [comment]
@zifgu The inconsistency https://github.com/tapestry-tool/tapestry-wp/pull/1222#issuecomment-1198784702 should be fixed with the latest commit.

As for https://github.com/tapestry-tool/tapestry-wp/pull/1222#issuecomment-1198843176, it no longer gives that warning and successfully deletes the node. I think it is fixed. However while testing, I noticed a bug with the sidebar where if you delete a node while having the sidebar open to that node, the sidebar will not close but instead render a blank area due to errors during the rendering (node data gone after deletion). I will fix it in the next commit.

image

**Comment by legendword** [comment]
The aforementioned bug https://github.com/tapestry-tool/tapestry-wp/pull/1222#issuecomment-1203220929 is fixed with the latest commit. It turns out that the issue was much more than just that.. the nodeId in the route remains the deleted node ID when node modal is closed after deletion, which is troublesome.

File: templates/vue/src/components/modals/NodeModal/index.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (zifgu) [reply templates/vue/src/components/modals/NodeModal/index.vue:739]: Sounds good, thanks for clarifying.

**Review by zifgu — APPROVED** [review]
Tested creating standalone draft nodes and draft nodes that are only connected to draft nodes. In the first case, an admin can directly publish the node, and in the second case, they cannot. The inconsistency https://github.com/tapestry-tool/tapestry-wp/pull/1222#issuecomment-1198784702 looks fixed, thanks!

I also tested deleting a node while the sidebar is still open; the selected node shifts to the parent if there is one, and otherwise goes to the default node. https://github.com/tapestry-tool/tapestry-wp/pull/1222#issuecomment-1203220929

The code looks good to me overall.

**Comment by wynnset** [comment]
> the available operations are different when a draft node created by an admin has a parent and when it is standalone

The admin should never have an option to submit for review. This should be fixed so that the option becomes "Save and Publish"

**Comment by wynnset** [comment]
Let's also create an update (using the `plugin-updates.php` file) that deletes any links that are not attached to any nodes as a part of this PR.

**Comment by legendword** [comment]
@zifgu To test the removal of invalid links, uncomment the following lines; the update will be run repeatedly since the plugin version has not been set to 3.0.0-beta yet. 
https://github.com/tapestry-tool/tapestry-wp/blob/98a195dd459596d2f9612c6ce3cbe1ee83df036f/plugin-updates.php#L92-L95

Be sure to also comment out the following line before you test the update, since once this is run the changes will be saved to database and you will no longer be able to see what it did. Instead add some error_log calls around this line to see which links are being removed.
https://github.com/tapestry-tool/tapestry-wp/blob/98a195dd459596d2f9612c6ce3cbe1ee83df036f/plugin-updates.php#L73

As for testing the "admin can publish any draft node", check the behaviour when an admin is trying to publish a draft node that is only connected to other draft nodes. Also for a normal user they shouldn't be able to directly publish any of their own draft nodes.

**Review by zifgu — COMMENTED** [review]
Tested the removal of invalid links, and links to nonexistent nodes are being filtered out correctly. Admins are now able to publish any draft nodes as well, while subscribers (almost) always need to submit for review.

I encountered one exception where subscribers can directly publish nodes. If subscribers should never be able to publish, then we may wish to fix this.
1. As a subscriber, create a draft node.
2. Edit the draft node and submit it for review.
3. As an administrator, accept the draft node.
4. As a subscriber, try to create a child of the accepted node. You will be allowed to directly publish it.

**Comment by zifgu** [comment]
The above bug (https://github.com/tapestry-tool/tapestry-wp/pull/1222#pullrequestreview-1066000658) only occurs  when the node is first saved as a private draft, then submitted for review, not when it is directly submitted for review. This is because an author-specific permission is added to nodes saved as private drafts, allowing them to pass Check 7 in `hasPermission` when all other checks fail.

The previous check only looks for review comments, which excludes "private draft" nodes without review comments. This seems like a bug to me; since users already get all permissions on draft nodes they author, by Check 2, we wouldn't need to add user-specific permissions while the node is still a draft.

https://github.com/tapestry-tool/tapestry-wp/blob/a20125c7a06859365cc118797ec22d5c7bb26219/classes/class.tapestry.php#L170-L175

We can fix this with the following change (done in caf4fe1):

https://github.com/tapestry-tool/tapestry-wp/blob/caf4fe18fc0e3cbc9d0d67c265debc483865a232/classes/class.tapestry.php#L172

I think this is reasonable, but I can't be sure that the previous check was not implemented for a reason. If you find something broken by this change, please let me know and I can try to fix it by changing `hasPermissions` instead.

**Comment by legendword** [comment]
I think the code change in https://github.com/tapestry-tool/tapestry-wp/pull/1222/commits/caf4fe18fc0e3cbc9d0d67c265debc483865a232 is reasonable and seems to work well. I was able to replicate the bug in another branch and wasn't able to reproduce it in this branch. 

Before ("1" is directly submitted for review and accepted, "B" is first saved as private draft, then submitted for review and accepted; editing, deleting, or directly publishing a child node to "B" is unrestricted to the subscriber author who should not have those permissions):

Pasted Graphic

(...continued: node "BB" is directly published by the author of "B1" who is only a subscriber and should not be able to directly publish any node)

Pasted Graphic 1

After (author no longer has permissions for the cases described above):

Pasted Graphic 2

**Comment by legendword** [comment]
Tested adding a draft node on an empty tapestry as admin. Other users will see the "The requested tapestry is empty" message as expected. Also tested creating nodes on an empty tapestry simultaneously, which correctly results in two standalone nodes.

**Comment by legendword** [comment]
I can confirm that non-admin users cannot move an accepted draft node; this used to be a bug and is now fixed.

**Comment by Bitajkr** [comment]
I am unable to delete the links between nodes in the following tapestry: https://test.tapestry-tool.com/1221-standalone-nodes/tapestry/testing-stand-alone-nodes/#/nodes/5?depth=0

**Comment by legendword** [comment]
Updated draft node permissions. The changes are recorded in the PR description.

**Comment by Bitajkr** [comment]
I just tested this out again today and it's working great! (https://test.tapestry-tool.com/1221-standalone-nodes/tapestry/testing-stand-alone-nodes-november-4-2022/#/nodes/16?x=275.9308&y=104.7595&scale=1.18)

Good Job!!

**Comment by Bitajkr** [comment]
**Bug:** Sometimes I am able to create a standalone node by using the authoring toolbar's node addition button; however, sometimes this feature doesn't work. In other words, the node either won't actually stick to the page when I try to add it, or the node that ends up being created isn't clickable and keeps moving around the page.

**Comment by Bitajkr** [comment]
### **Bug Report:** 
I tried adding a standalone node by following the following step
1. vfbfbvfg
2. f fgbfg 
3. vfbf f
4. On this step, my node crashed. 

**Expected Behaviour:** 
I expected that when I press publish, my node would show up right away. 

**Actual Behaviour:** 
When I press publish, the node crashes or doesn't open (explain what this mean) 

MUST ADD PICTURE
